flax/linen/summary.py

```diff
@@ -0,0 +1,435 @@
+
+import dataclasses
+import io
+from typing import Any, Callable, Dict, Iterable, List, Mapping, Optional, Sequence, Set, Tuple, Union
+
+import flax
+import jax
+import rich.table
+import rich.text
+import yaml
+from flax.core.scope import CollectionFilter, DenyList
+
+PRNGKey = Any  # pylint: disable=invalid-name
+RNGSequences = Dict[str, PRNGKey]
+Array = Any    # pylint: disable=invalid-name
+
+
+@dataclasses.dataclass
+class Row:
+  """Contains the information about a single row in the summary table.
+
+  Attributes:
+    path: A tuple of strings that represent the path to the module.
+    outputs: Output of the Module as reported by `capture_intermediates`.
+    module_variables: Dictionary of variables in the module (no submodules included).
+    counted_variables: Dictionary of variables that should be counted for this row,
+      if no summarization is done (e.g. `depth=None` in `module_summary`) then this field
+      is the same as `module_variables`, however if a summarization is done then this
+      dictionary potentially contains parameters from submodules depending on the depth
+      of the Module in question.
+  """
+  path: Tuple[str, ...]
+  outputs: Any
+  module_variables: Dict[str, Dict[str, Array]] # variables of the row's corresponding module
+  counted_variables: Dict[str, Dict[str, Array]] # all variables that should be counted for this row
+
+  def size_and_bytes(self, collections: Iterable[str]) -> Dict[str, Tuple[int, int]]:
+    return {
+      col: _size_and_bytes(self.counted_variables[col])
+      if col in self.counted_variables else (0, 0)
+      for col in collections
+    }
+
+class Table(List[Row]):
+  """A list of Row objects.
+
+  Table inherits from `List[Row]` so it has all the methods of a list, however
+  it also contains some additional fields:
+  
+  * `module`: the module that this table is summarizing
+  * `collections`: a list containing the parameter collections (e.g. "params", "batch_stats", etc)
+  """
+  def __init__(self, module: "flax.linen.Module", collections: List[str], rows: Iterable[Row]):
+    super().__init__(rows)
+    self.module = module
+    self.collections = collections
+
+
+def module_summary(
```

Comment:
The name seems to suggests this will return a summary, but it actually returns a Table object, which isn't very useful, so I wonder if it is useful to expose this function at all. 

Alternatively you could add a function `tabulate` to this file that calls `module_summary` (I would rename it to `_get_module_table`) and `render_table` (I would rename it to `_render_table`).

Otherwise, users may try to call `summary.module_summary` expecting a summary of the model.

Reply:
Plan:
- rename `module_summary` => `_get_module_table`
- rename `render_table` => `_render_table`
- create `summary.tabulate` that follows @jheek transformation style suggestion.